src/query/builder.js

```diff
@@ -57,9 +57,16 @@ assign(Builder.prototype, {
     return cloned;
   },
 
-  timeout(ms) {
+  timeout(ms, {cancel} = {}) {
     if(isNumber(ms) && ms > 0) {
       this._timeout = ms;
+      if (cancel) {
+        if (this.client.canCancelQuery) {
+          this._cancelOnTimeout = true;
+        } else {
+          throw new Error("Query cancelling not supported for this dialect");
```

Comment:
Would it be possible to make the client-type-agnostic part of knex perform this check by having it provide a wrapper `timeout()` function that then calls this function only if the check passes? That would kill the duplication and still give us the immediate error feedback.

Haven't checked the relevant knex code though, so might me talking out of my ass here. :smile: